Gets detailed information about specified user-id

anil-kumbhar opened the pull request:
According to Identity API Specification [1], user can get details about herself with following request:

GET v2.0/users/<USER_ID>

This was not supported by SoftLayer driver, we are supporting this using SoftLayer_Account service's getCurrentUser [2] method.

[1] http://api.openstack.org/api-ref-identity.html#identity-v2
[2] http://sldn.softlayer.com/reference/services/SoftLayer_Account/getCurrentUser

anil-kumbhar committed:
Gets detailed information about specified user-id

According to Identity API Specification [1], user can get details about herself with following request:

GET v2.0/users/<USER_ID>

This was not supported by SoftLayer driver, we are supporting this using SoftLayer_Account service's getCurrentUser [2] method.

[1] http://api.openstack.org/api-ref-identity.html#identity-v2
[2] http://sldn.softlayer.com/reference/services/SoftLayer_Account/getCurrentUser

beittenc commented:
Looks like the Travis build failed due to a checkout issue on their side. Could you force the build to run again? I expect the tests will pass.

anil-kumbhar commented:
Not able to force the build to run again. Tried most of the solutions mentioned here [1], however, interface doesn't have an option for me. Since, you might have write access to this repository can you rebuild? Thanks!

[1] http://stackoverflow.com/questions/17606874/trigger-a-travis-ci-rebuild-without-pushing-a-commit

beittenc commented:
I also don't have the normal force rebuild options. I guess just make a minor commit and force it that way.

sudorandom commented:
I restarted the build.

bodenr commented:
if I understand this patch set, you are effectively implemented the following keystone API:

```
GET v2.0/users/​{user_id}​
```

a few considerations / concerns about the impl in this patch:
- i believe this API is mainly intended for use with an 'admin token' to retrieve details about some other user specified by their user id. the way its currently coded it appears to totally ignore the `user_id` request parameter and return details about the current user (as per retrieved data with SL client using the tokens SL api_key).
- as mentioned above, I don't see any checking or use the of `user_id` request parameter. so if my user id is `boden` (as per my api_key which is transported in the jumpgate token) I can call it like `GET /v2.0/users/kevin` and it'll give me back the user details for `boden` still... seems semantically incorrect / confusing.

to me this impl might consider:
- if the api_key (in the X-Auth-Token) is for an SL account owner, it should actually get the details for the user_id specified in the request parameter. that is -- if you are a SL account owner and can prove such (via your api_key transported in your token), then you can get details about other users in your account.
- if the api_key is only an account user, then the method is only applicable when the user_id request parameter is for the api_key user.. otherwise its an error.

sudorandom commented:
This logic might be more appropriate for a new manager in the softlayer-python package itself but here's code that is able to find a user by username if the user has appropriate permissions:

``` python
sl_client['Account'].getUsers(filter={'users': {'username': {'operation': 'your_username'}}})
```

You'd probably want to use a mask to only bring back the data that this call requires.

anil-kumbhar committed:
Retrieved specified user details.

Corrected getCurrentUser with getUsers with specified user id.

anil-kumbhar commented:
corrected suggested changes. Now it'll give user details of specified user-id.

sudorandom commented on jumpgate/identity/drivers/sl/user.py:
```diff
@@ -0,0 +1,20 @@
+        user = client['Account'].getUsers(
+            mask='mask[id,username,accountId]', filter={'users': {'id': {'operation': user_id}}})
+        if not user or not len(user):
+            raise Exception("User with ID '%s' Doesn't Exists" % user_id)
```
One last nitpic: This should probably return after setting the response to not found (404) instead of raising a generic exception. Here's an example of that: https://github.com/softlayer/jumpgate/blob/master/jumpgate/compute/drivers/sl/servers.py#L39

amol replied:
Does not getting user for specified ID translates to 404, can it also be 403 (Unauthorized)? If so how can we distinguish?

sudorandom replied:
I'm not sure what permission controls the ability to introspect users on the same account but we'd have to check that to determine if it's a permissions thing or the user just doesn't exist. I'm okay with defaulting it to 404 until that's figured out.

Commit:
Merge branch 'user-get-request' of https://github.com/anil-kumbhar/jumpgate into user-get-request

Commit:
If user doesn't exists corresponding HTTP code/message has been given

bodenr commented on jumpgate/identity/drivers/sl/user.py:
```diff
@@ -0,0 +1,18 @@
+        if not user or not len(user):
+            return not_found(resp, "Invalid user ID specified.")
+        user_response = {'id': str(user[0]['id']),
+                         'username': str(user[0]['username']),
```
as per keystone identity v2 api here http://api.openstack.org/api-ref-identity.html#identity-v2 shouldn't the username be returned as the key `name` not as `username`?

Here's the sample response from the openstack api docs, notice the `"name":"jqsmith"`

```
{
   "user":{
      "id":"u1000",
      "name":"jqsmith",
      "email":"john.smith@example.org",
      "enabled":true
   }
}
```

bodenr commented on jumpgate/identity/drivers/sl/user.py:
```diff
@@ -0,0 +1,18 @@
+                         'username': str(user[0]['username']),
+                         'email': str(user[0]['username']),
+                         'tenantId': str(user[0]['accountId']),
+                         'enabled': 'true'
```
in the JSON response over the wire, does this get translated into a JSON boolean or is it a string over the wire? I think it will get marshalled into a string...

It needs to be a boolean -- not a string... 

See sample python session and notice the difference between `true` and `'true'` -- the former is a JSON boolean and the later is a string.

``` python
Python 2.7.3 (default, Feb 27 2014, 19:58:35)
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import json
>>> json.dumps({'enabled':'true'})
'{"enabled": "true"}'
>>> json.dumps({'enabled':True})
'{"enabled": true}'
```

bodenr commented on jumpgate/identity/drivers/sl/user.py:
```diff
@@ -0,0 +1,18 @@
+
+    def on_get(self, req, resp, user_id):
+        client = req.env['sl_client']
+        user = client['Account'].getUsers(
```
so what happens here if the callers request token (api_key) does not have auth to view other users in the account?? do we end up with an exception from the `getUsers()` call which results in a 500 response to the caller?

It seems to me we want to explicitly handle such a case and instead return a 401... That is, if the caller does not have the authority to view other users the result is a 401 (unauthorized), not a 500.

Commit:
If user is unauthorized corresponding HTTP code/message has been given

Commit:
Corrected pep8 error from previous build

anil-kumbhar commented:
In the response there is also a "username" property that openstack will display along with "name", so no need to replace username by name. 
We can retrieve "name" and "email" property only for child users. For getting child user's email
and name we can use SL service SoftLayer_User_Customer::getChildUser(mask="mask[email,firstName,lastName]",id=147684), 
where id is the parent user's id and the parent has appropriate rights to get child's data. And we can get current user's parent ID. What you suggest on this approach?

Commit:
User can display details of itself and its child only

Commit:
Corrected pep8 error from previous build

sudorandom commented:
Woah, this one got a little too complicated. From my testing it looks like a permission-limited user can only see subusers using the call: `client['User_Customer'].getObject(id=12345, mask='id,accountId,username,firstName,lastName,email')`. If either the user does not exist or you don't have permission to view the user a `SoftLayer.SoftLayerAPIError` exception is thrown with e.faultCode of `SoftLayer_Exception_ObjectNotFound`. I believe treating this as a 404 and not trying to distinguish a 403 is the best decision. If you were able to determine which user ids are valid by attempting this call with every user id would be a leak of data and I feel like trying to make this determination through other means will lead to the implementation being fragile.

Commit:
Merge branch 'user-get-request' of https://github.com/anil-kumbhar/jumpgate into user-get-request

anil-kumbhar committed:
Always raised 404, even if user is unauthorized

User can retrieve details of all users that he has access to otherwise 404 HTTP error will be returned.

sudorandom commented on jumpgate/identity/drivers/sl/user.py:
```diff
@@ -0,0 +1,42 @@
+            kwargs['mask'] = "mask[%s]" % ','.join(items)
+
+        client = req.env['sl_client']
+        current_user = client['Account'].getCurrentUser(**kwargs)
```
I believe this call can be removed entirely.

sudorandom commented on jumpgate/identity/drivers/sl/user.py:
```diff
@@ -0,0 +1,42 @@
+        try:
+            user = client['User_Customer'].getObject(id=user_id,
+                                                     **kwargs)
+        except Exception:
```
I'd prefer stricter exception checking, like the following:

``` python
except SoftLayer.SoftLayerAPIError as ex:
    if ex.faultCode == 'SoftLayer_Exception_ObjectNotFound':
        return not_found(resp, "Invalid User ID specified")
    raise
```

Commit:
Merge pull request #74 from sudorandom/json-extention

Adds .json suffix to each route

Commit:
Merge branch 'user-get-request' of https://github.com/anil-kumbhar/jumpgate into user-get-request

anil-kumbhar committed:
Handled appropriate exception and raised others

Apart from exception handling user details are given if present.

anil-kumbhar commented:
For some users when it tries to retrieve details of it's parents ( Upper level users of a child) then it gets only  "id", "accountId", "username" fields. Remaining fileds "name" and "email" doesn't get. I think in this case jumpgate should return available data about given user-id.

anil-kumbhar commented:
Can this be merged?

sudorandom commented:
This LGTM. I'll let @beittenc review/merge this one.

anil-kumbhar commented:
@beittenc: Can this be merged? Or any suggestions from your side on this?

beittenc commented on jumpgate/identity/drivers/sl/user.py:
```diff
@@ -0,0 +1,42 @@
+            user = client['User_Customer'].getObject(id=user_id,
+                                                     **kwargs)
+        except SoftLayerAPIError as ex:
+            if ex.faultCode == 'SoftLayer_Exception_ObjectNotFound':
```
This results in a data leak. If we return a 404 for an invalid user ID and a different error for a valid user ID, Jumpgate can be used to verify an account's existence as part of a general attack strategy. All authentication failures should be returned with a consistent error.

sudorandom replied:
There's no data leak. It's a 404 if either the user does not actually exist or you don't have access to view the user. Both result in a SoftLayer_Exception_ObjectNotFound error. All other types of errors (unauthed, server error, etc) are re-raised by the raise statement below.
